### DISCUSSION TITLE
Using `with_loader_criteria` in nested relationship chain

### DISCUSSION THREAD

**Discussion by igoras1993** [sqlalchemy/sqlalchemy#8711] [Usage Questions] [answered]
Hi,

I know that this is an old thread but I would like to elaborate more on #5053 in context of `1.4`. @zzzeek please correct me if I am wrong, but `1.4` release added a [`with_loader_criteria`](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.with_loader_criteria) option. It can be used for applying criteria on arbitrary relationship, right?

```python3
stmt = select(A).options(
  selectinload(A.bs)
  with_loader_criteria(B, B.some_column > 10)
)
```

It should limit eagerly loaded `A.bs` to only such `B` that `B.some_column` is greater than `10`, right?

In such context, I want to ask another question. Is there a way to apply such options in a nested manner? Suppose something like:

```python3
stmt = select(A).options(
  selectinload(A.bs).options(
    selectinload(B.cs),
    with_loader_criteria(C, C.amount > 10)
  )
)
```

This raises an error:
```
AttributeError: 'LoaderCriteriaOption' object has no attribute '_apply_to_parent'
```

It can be "fixed" by placing `with_loader_criteria` in the top-level options:

```python3
stmt = select(A).options(
  with_loader_criteria(C, C.amount > 10),
  selectinload(A.bs).options(selectinload(B.cs))
)
```

This "fix" works as long as there is no other relationship to `C` in a relationship chain and no distinct criteria is needed for that other relationship. Is there a way for using `with_loader_criteria` on particular relationship loading options, not "globally"?

**Comment by zzzeek** [comment, marked as answer]
that's the expected behavior (other than the not-good exception message, which should be fixed).  with_loader_criteria() is a query-global option that applies to entities based on what class they are, and is not specific to "paths".

**Reply by igoras1993** [reply]
Thanks for reply. I've assumed that this is intended. Are there any plans on allowing such "filtering" to specific paths in future releases? It would be helpful to have such an option in sqlalchemy arsenal.

**Reply by zzzeek** [reply]
yes you can do this right now.   See [Adding criteria to loader options](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#adding-criteria-to-loader-options).

**Reply by igoras1993** [reply]
Wow, thank You very much. Exactly what I was looking for. Didn't know that it can be applied to `selectin` strategy. I've just tried it out, looking at resulting SQL and it works flawlessly. Thank You very much again!

**Comment by jlynchMicron** [comment]
Hi @zzzeek, 

I have been struggling with trying to make this idea work using selectinload and with_loader_criteria for deeply nested relationships where I dont want the high level object loaded if it does not meet low level object filter criteria.  The reason why I am using selectinload is because I am using SQLite with more than 64 tables used in the full high level object relationship and SQLite cannot join on more than 64 tables.  The selectinload works fine for constructing the high level object if not with_loader_criteria filtering is used, but when I use with_loader_criteria filtering, it will just nullify the value of the low level object attribute in the high level object, whereas my desired behavior is stopping that particular high level object from being loaded if it does not meet low level filter criteria.  Is there any way of achieving this behavior with selectinload + with_loader_criteria, a way to break up joinedload (dynamically, not hard coded), and/or a good general recursive loader for very large relationship structures?

I have been using this function I made for recursive loading, but its low level object filtering does not seem to be working correctly (as described above):
```
def orm_recursive_loader(
        orm_dclass:type, 
        loading_technique:_AbstractLoad = sa_orm.selectinload,
        #loading_technique:_AbstractLoad = partial(sa_orm.joinedload, innerjoin=True), #Pass callable with argument set.
        loader_criteria:List[Tuple[type, _ColumnExpressionArgument]] = [],
    ) -> List[_AbstractLoad]:
    """
    Recursive function that creates ORM loading behavior similar to joinedload('*') via other loading techniques.
    This is handy for breaking apart large single joinedload operations into a set of smaller operations.
    Further documentation: https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#summary-of-relationship-loading-styles

    Example:
        Assuming you have Grandparent, Parent, ChildA, and ChildB ORM classes defined; this function produces similar options to this code.
        query = session.query(Grandparent).options(
            selectinload(Grandparent.parents).options(
                selectinload(Parent.children_a),
                selectinload(Parent.children_b)
            )
        )

    Additional loader criteria can be applied via the loader_criteria argument, which essentially applies additional WHERE statements to further filter what is loaded.
    Further documentation: https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria

    Example:
        stmt = select(User).options(
            selectinload(User.addresses),
            with_loader_criteria(Address, Address.email_address != 'foo'))
        )

    Args:
        orm_dclass (type): SQLAlchemy ORM dataclass 
        loading_technique (_AbstractLoad, optional): Desired SQLalchemy loading technique to use between table relationships. Defaults to sa_orm.selectinload.
        loader_criteria (List[Tuple[type, _ColumnExpressionArgument]], optional): Additional WHERE criteria to apply to the loading operation. Defaults to empty list.

    Returns:
        List[_AbstractLoad]
    """
    loader_options = []

    for f in fields(orm_dclass):
        if is_dataclass(f.type):
            dtype = f.type
        elif get_origin(f.type) is list and is_dataclass(get_args(get_type_hints(orm_dclass)[f.name])[0]):
            dtype = get_args(get_type_hints(orm_dclass)[f.name])[0]
        else:
            continue

        #Add any dataclass field to loader options.
        loader = loading_technique(orm_dclass.__dict__[f.name])

        #If dataclass field also had dataclass fields, add those as well.
        child_loader_options = orm_recursive_loader(dtype)
        if child_loader_options:
            loader = loader.options(*child_loader_options)
        
        loader_options.append(loader)

    if loader_criteria:
        for criteria_type, criteria_arg in loader_criteria:
            loader_options.append(sa_orm.with_loader_criteria(criteria_type, criteria_arg))

    return loader_options
```

This function's results will be passed into my query options.

**Simply put, how do I negate the loading big/complex high level objects if they do not meet the criteria placed on one of their associated low level objects?**

**Reply by zzzeek** [reply]
this would be a new discussion but the short answer is: think in SQL.   I want to SELECT some rows but only if they meet some criteria on a related table.       Then I want that related table criteria all inside the WHERE clause.   So the answer is always, use a correlated subquery:

```
SELECT * FROM high_level_table WHERE EXISTS (SELECT * from low_level_table WHERE my_criteria AND low_level_table.id = intermediary_table.some_id AND intermediary_table.some_id = high_level_table.some_id)
```

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import exists
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
from sqlalchemy.orm import with_loader_criteria


class Base(DeclarativeBase):
    pass


class TopLevel(Base):
    __tablename__ = "toplevel"

    id: Mapped[int] = mapped_column(primary_key=True)

    a_collection = relationship("A")


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    toplevel_id = mapped_column(ForeignKey("toplevel.id"))

    data: Mapped[str]
    b: Mapped[list[B]] = relationship("B")
    b_id: Mapped[int] = mapped_column(ForeignKey("b.id"))

    def __repr__(self):
        return f"A(data={self.data})"


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(
    TopLevel(
        a_collection=[
            A(data="we want 1", b=B(data="we want")),
            A(data="we dont want 2", b=B(data="we dont want")),
            A(data="we dont want 3", b=B(data="we dont want")),
            A(data="we want 4", b=B(data="we want")),
        ]
    )
)
s.commit()

# an exists query (the pep484 typing not quite accepting that WHERE as written...)
stmt = select(TopLevel).options(
    with_loader_criteria(A, exists(B).where(B.data == "we want").where(A.b))
)

# or just use has() (same thing)
stmt = select(TopLevel).options(
    with_loader_criteria(A, A.b.has(B.data == "we want"))
)

for tl in s.scalars(stmt):
    print(tl.a_collection)

```

**Reply by jlynchMicron** [reply]
Thanks for your reply! Would this method work for deeply nested relationships?  What if the TopLevel object was a few more relationship hops away from A and B?  If A.b did not meet the loader criteria, would it cause the TopLevel object not to be loaded or would it just appear that the mid-relationship object with a relationship to A having a Null attribute of A (since its underlying A.b did not meet loader criteria)?  I am trying to find a way of loading deeply nested classes with criteria, ideally without having to put relationship criteria at every level between the top level object and low level object when filtering on low level objects.

The only way I could see this working is putting in the entire relation chain in front of the "has" criteria. 
Example: with_loader_criteria(A, A.b.c.d.e.f.has(F.data == "we want"))

**Reply by zzzeek** [reply]
it will work for any level of nesting as long as you can build a SQL expression that hops across all the elements to be joined.  This is not something that can be generalized from A->B without explicitly noting the elements that connect A and B.  Also, correlated subqueries as they get more nested are going to perform more and more poorly.

to do the actual nesting, you need to nest explicitly, mappings as:

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    toplevel_id = mapped_column(ForeignKey("toplevel.id"))

    data: Mapped[str]
    b: Mapped[list[B]] = relationship("B")
    b_id: Mapped[int] = mapped_column(ForeignKey("b.id"))

    def __repr__(self):
        return f"A(data={self.data})"


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)

    c: Mapped[list[B]] = relationship("C")
    c_id: Mapped[int] = mapped_column(ForeignKey("c.id"))

class C(Base):
    __tablename__ = "c"
    id: Mapped[int] = mapped_column(primary_key=True)

    d: Mapped[list[B]] = relationship("D")
    d_id: Mapped[int] = mapped_column(ForeignKey("d.id"))

class D(Base):
    __tablename__ = "d"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

```

nesting is:

```py
s.add(
    TopLevel(
        a_collection=[
            A(data="we want 1", b=B(c=C(d=D(data="we want")))),
            A(data="we dont want 2", b=B(c=C(d=D(data="we dont want")))),
            A(data="we dont want 3", b=B(c=C(d=D(data="we dont want")))),
            A(data="we want 4", b=B(c=C(d=D(data="we want")))),
        ]
    )
)
s.commit()

stmt = select(TopLevel).options(
    with_loader_criteria(A, A.b.has(B.c.has(C.d.has(D.data == "we want"))))
)

```

this produces SQL that might not scale:

```
SELECT a.id AS a_id, a.data AS a_data, a.toplevel_id AS a_toplevel_id, a.b_id AS a_b_id 
FROM a 
WHERE ? = a.toplevel_id AND (EXISTS (SELECT 1 
FROM b 
WHERE b.id = a.b_id AND (EXISTS (SELECT 1 
FROM c 
WHERE c.id = b.c_id AND (EXISTS (SELECT 1 
FROM d 
WHERE d.id = c.d_id AND d.data = ?))))))

```


If I had this problem to solve in an actual production application I'd be looking to get a column onto "toplevel" here that materializes the value we are seeking from the nested objects (i.e. denormalizes this value ahead of time for fast reads).   This column could be populated by the application using events, or by triggers written in the database, or perhaps even using a computed column.

**Reply by jlynchMicron** [reply]
Thanks again for the advice!  Yeah, unfortunately my application is for some old data that was never really intended to be stored in a database leading to some nasty relation chains.  Luckily we are using GCP bigquery as the database engine so hopefully its native platform scalability with overcome poor schema design 😄

**Reply by zzzeek** [reply]
here's a much better query

```py
subq = select(1).select_from(B).where(A.b).join(B.c).join(C.d).where(D.data == "we want").exists()

stmt = select(TopLevel).options(
    with_loader_criteria(A, subq)
)

```

output

```sql
SELECT a.id AS a_id, a.data AS a_data, a.toplevel_id AS a_toplevel_id, a.b_id AS a_b_id 
FROM a 
WHERE ? = a.toplevel_id AND (EXISTS (SELECT 1 
FROM b JOIN c ON c.id = b.c_id JOIN d ON d.id = c.d_id 
WHERE b.id = a.b_id AND d.data = ?))

```

**Reply by CaselIT** [reply]
> perhaps even using a computed column.

AFAIK it's not possible to use other tables in a computed column. It must be derivable from the table row

> Luckily we are using GCP bigquery as the database engine so hopefully its native platform scalability with overcome poor schema design 

Careful that big query can scale a lot but it's not doing that for free :)

edit: you may also consider a (materialized) view that pre-joins the data, so that you would simply apply filters on columns. I'm not sure how feasible it is in your case since you seem to have 64+ long chains but maybe you can fine compromises to reduce the number of nesting you have to do

**Reply by jlynchMicron** [reply]
I am well aware of cost overruns in BigQuery from personal experience 😅.  
We use both materialized views and reserved compute slot cost structuring to help avoid query cost overruns.

Also for clarity, my relation chain is not +64 tables deep, but ~90 tables of varying relation depth depending on the relation path from the top level object.  We basically took old program data, wrote python dataclasses to represent the component parts, then made an ORM from said dataclasses.  Unfortunately I could not use joinedload('*') for SQLite testing of the ORM since it only supports up to 64 tables in a single query.

**Reply by jlynchMicron** [reply]
Hi @zzzeek and @CaselIT, as a follow up to our conversation, I am considering making a WHERE statement query optimizer for my tree data structure where a user can put in various (node.attr {conditional} value) statements and my function will generate the most optimal WHERE statement via analyzing the relation graph.  

Out of pure curiosity, does the underlying mechanics of ORM query evaluation work in a similar graph-based manner?  I know many programming compilers (like LLVM I think) rely on graph-based approaches for compilation, and I was wondering if SQLalchemy did the same.  If SQLalchemy does do this, I wonder if I can build my relation analyzer in a more SQLalchemy "native" way that could maybe be shared with others.  Otherwise, I will probably just build a relation graph with a graph library like "networkx" and just analyze the graph that way.

Simple example:
 - Relation list: A - B, B - C. B - D
 - Query criteria: C.id = 1, D.id =50
 - Resultant WHERE statement pseudocode: 
``` 
A.b_relation.has(
    sa._and(
        B.c_relation.has(
            C.id == 1
        ),
        B.d_relation.has(
            D.id == 50
        )
    )
)
```

**Reply by zzzeek** [reply]
the ORM has a query "evaluator" that is a component of the ORM-enabled DDL functions like bulk UPDATE, but I think more generally you're referring to compilation of queries as well as how it builds up queries.   SQL expressions like all language syntaxes are represented by parse trees and SQLAlchemy is no different.   Analyzing and/or augmenting a SQL query is a graph traversal problem like any other, so yes the ORM relies entirely on lots of graph traversal to do things, not just generating SQL strings but also detecting cartesian products, determining how tables should be joined together and if that implies one-to-many or many-to-one, and lots of other decisions of that nature.
